File: 22-generate-parentheses/generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def REC(op,cl,arr):
            if op == n and cl == n:
                result.append("".join(arr))
                return
            if op < n:
                arr.append('(')
                REC(op+1,cl,arr)
                arr.pop()
            if cl < op:
                arr.append(')')
                REC(op,cl+1,arr)
                arr.pop()
        result = []
        REC(0,0,[])
        return result

File: 22-generate-parentheses/generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> arr = new ArrayList<>();
        StringBuilder sb = new StringBuilder();
        rec(sb,0,0,n,arr);
        return arr;
    }
    public static void rec(StringBuilder res,int open,int close,int n,List<String> result){
        if(res.length() == 2*n){
            result.add(res.toString());
            return;
        }
        if(open < n){
            res.append("(");
            rec(res,open+1,close,n,result);
            res.deleteCharAt(res.length()-1);
        }
        if(close < open){
            res.append(")");
            rec(res,open,close+1,n,result);
            res.deleteCharAt(res.length()-1);
        }
    }
}